### ISSUE TITLE
pachctl logs should observe access controls

### ISSUE THREAD

**Issue by pappasilenus** [pachyderm/pachyderm#6317]
**What happened?**:

A user with access "none" for a pipeline's output repo was able to see logs related to that pipeline, mentioning file names and sizes, from the main pachd logs when running `pachctl logs`.

**What you expected to happen?**:

I expected the logs to be filtered by the user's access level.

**How to reproduce it (as minimally and precisely as possible)?**:

Prerequisite: A pachyderm workspace with access controls enabled and 2 users, call them Wilma and Betty.

1. Wilma creates an input repo and a pipeline that uses it as input.  images/edges from OpenCV is adequate.
2. Wilma assigns Betty access "none" : 
```
pachctl auth set Betty none images
pachctl auth set Betty none edges
```
3. Wilma adds some data to images that triggers action in edges.
4. Wilma runs some commands on edges, like `pachctl list file edges@master`
5. Betty runs the command `pachctl logs`.  Log lines such as this are visible to her:
```
2021-06-09T22:50:22Z INFO auth.API.Authorize {"duration":0.000491647,"request":{"repo":"images","scope":1},"response":{"authorized":true}} []
2021-06-09T22:50:22Z INFO pfs.API.InspectCommit {"duration":0.003208283,"request":{"commit":{"repo":{"name":"images"},"id":"cf9e9a99707242bc86b6fdb4a34e3232"}},"response":{"commit":{"repo":{"name":"images"},"id":"cf9e9a99707242bc86b6fdb4a34e3232"},"branch":{"repo":{"name":"images"},"name":"master"},"origin":{},"parent_commit":{"repo":{"name":"images"},"id":"7f734142cbe2442fbe2fd67f750c3167"},"started":{"seconds":1623278428,"nanos":903761612},"finished":{"seconds":1623278429,"nanos":562884625},"size_bytes":244068,"subvenance":[{"lower":{"repo":{"name":"edges"},"id":"525490b2789f46e6be9d9d493198c772"},"upper":{"repo":{"name":"edges"},"id":"525490b2789f46e6be9d9d493198c772"}},{"lower":{"repo":{"name":"edges"},"id":"f7c5f6fdecab40bbae096a38bdd71a23"},"upper":{"repo":{"name":"edges"},"id":"f7c5f6fdecab40bbae096a38bdd71a23"}},{"lower":{"repo":{"name":"montage"},"id":"02959c1e5ecd41328a859d17842fab81"},"upper":{"repo":{"name":"montage"},"id":"02959c1e5ecd41328a859d17842fab81"}},{"lower":{"repo":{"name":"montage"},"id":"1f0cb4d7867a4c50a56a669edffa38c6"},"upper":{"repo":{"name":"montage"},"id":"1f0cb4d7867a4c50a56a669edffa38c6"}}],"tree":{"hash":"ebe16b5026ad41fe485ec4f6523198d8245822e25faaa33e29583242c6ec34077df1939a5ffc6fff3d21bfbfb48c41f734880f33f38c5e3513aee4f00dc163cf"},"subvenant_commits_success":4,"subvenant_commits_total":4}} []
2021-06-09T22:50:22Z INFO pfs.API.InspectFile {"request":{"file":{"commit":{"repo":{"name":"images"},"id":"cf9e9a99707242bc86b6fdb4a34e3232"},"path":"/46Q8nDz.jpg"}}} []
2021-06-09T22:50:22Z INFO pfs.API.InspectFile {"request":{"file":{"commit":{"repo":{"name":"images"},"id":"cf9e9a99707242bc86b6fdb4a34e3232"},"path":"/g2QnNqa.jpg"}}} []
2021-06-09T22:50:22Z INFO pfs.API.InspectFile {"request":{"file":{"commit":{"repo":{"name":"images"},"id":"cf9e9a99707242bc86b6fdb4a34e3232"},"path":"/8MN9Kg0.jpg"}}} []
```

**Anything else we need to know?**:

**Environment?**:
- Kubernetes version (use `kubectl version`):
- Pachyderm CLI and pachd server version (use `pachctl version`):
- Cloud provider (e.g. aws, azure, gke) or local deployment (e.g. minikube vs dockerized k8s):
- OS (e.g. from /etc/os-release):
- Others:

**Comment by actgardner** [comment]
I agree that we shouldn't be leaking repo/file names in the logs. There's a few possible solutions here, in ascending order of effort:

- Limit access to the pachd logs in pachctl. It seems like we benefit from these very verbose logs when debugging customer clusters (do we actually?), but we could restrict access to cluster admins (or a specific permission in 2.0).

- Remove the request/response bodies from the logs altogether. Logging the entire request and response is pretty unusual, and it's caused a lot of issues in the past by logging secrets, etc. We could move to a generic interceptor and clean up a lot of boilerplate this way as well.

- Filter the log lines by access controls. We could remove log lines if they refer to repos a user can't access. This would require some structured logging/parsing/processing which we don't currently have - it would be hard to preserve the existing format and support this with 100% accuracy. It also risks making our debug info useless - if a user has partial access they wouldn't be able to see a complete picture of cluster activity.

- Redact request/response by access controls. We could log every request/response but redact sensitive fields according to ACLs. This would require structured logging and a lot of engineering effort, but it would be nifty.

**Comment by pappasilenus** [comment]
My thoughts are  that a combination of the first and second option may be preferred, 

1. provide access controls on the pachd logs for the cluster-admin role
2. only do request/response for DEBUG logging level, which has to be explicitly set. Normal INFO logging level doesn't show request/response. 

If request/response information would be part of `pachctl debug dump`, that automatically puts debug logs in the dump, and requires cluster-admin to run, I think we're covered in terms of what support might need and a  rational, secure process around it. I've asked CS for input.

**Comment by jrockway** [comment]
How possible would it be to store logs in pfs instead of an external system (k8s/loki)?  It would be nice if the logs for a given pipeline ended up stored next to the outputs for that pipeline (image -> edges pipline logs end up in edges.logs for example); that way access checks could be done cleanly.  The standard logs printed to stdout could just contain references: `log("foo", {"object": "is here"})` would print "INFO foo edges.log.abc12321cba@1234" and the full details could be retrieved from the pfs file edges.log.abc12321cba at commit 1234.  (Would also be nice to store trace spans here, so that we don't have to send them to Jaeger.)

I worry that pfs doesn't do well with deduplicating data like this, so storage requirements might be large.  There is also a bootstrapping problem -- where do you log "hey, that s3 bucket doesn't exist" ;)

**Comment by mindthevirt** [comment]
Why don't we make pachctl logs, without arguments an admin command and support -p, -j, -? based on acls.
A user should be able to debug  any component without having to use pachctl logs without any flag. Or we could make pachctl logs require an argument and only allow no argument when run as part of debug dump.

**Comment by actgardner** [comment]
@mindthevirt That sounds like (1) - having a role for getting the pachd logs, which is separate from the permissions for pipeline/job logs? The CLI doesn't have to change, we'll just add an auth check.

> How possible would it be to store logs in pfs instead of an external system (k8s/loki)?

I think we put the pipeline logs in the stats commit already, but I don't think you can read from a commit in-progress to debug a system, or tail a commit, etc. This also wouldn't resolve this issue, because any time a user interacts with a repo by making an RPC to pachd it produces these logs.

**Comment by jrockway** [comment]
Why does `pachctl logs` ever need to return the main pachd logs?  Users need logs to debug their jobs, but that doesn't require pachd's own RPC logs.  (I think anyone should be able to read the logs from jobs they submit, without asking the admin for additional permissions.  But it's not actually clear to me; does submitting a pipeline that reads from "x" and "y" require the user that submitted that pipeline to also be able to read from "x" and "y"?  If not, they can break the ACL by just submitting a job to print /pfs/in.  It's orthogonal to logging, though -- they could also email themselves the files or something.)

That doesn't solve all problems -- logging secrets still potentially violates the contract (there are separate k8s permissions for "can read logs" and "can read secrets").  But I think it's correct to get pachd logs with `kubectl logs` and not `pachctl logs`.  

(Loki complicates this.  The logs all go into a pool in a way that's not aware of pach auth rules, and pachd is fully-trusted to make the right request within the scope of the accessing user.  Maybe it does that, but other workloads on the k8s cluster can probably write to Loki, and `pachctl logs` can't tell those logs from real logs, causing it to leak information.  The only thing to do there is to encrypt logs, and decrypt them when fetching them -- if they aren't encrypted, they're definitely not logs that pachd should stream back.  But... I bet nobody really cares about this.)

**Comment by actgardner** [comment]
> Why does pachctl logs ever need to return the main pachd logs? 

Yeah I wonder this as well - there's value in aggregating the logs from the individual pipelines for the user, and enforcing the access controls, but any k8s user who is administering pach can get the pachd logs.

**Comment by actgardner** [comment]
This should be resolved for 2.0 - we have a new `pachdLogsReader` role which can be used to give users access to pachd logs. Changing the format of the logs is potentially a future change but it's out of scope for 2.0
